### PR TITLE
Clean up routes module slightly

### PR DESCRIPTION
**Pull request by jD91mZM2**
- Replace update with get_ref/get_mut
- Rename ipv4 add_default_ipv4_route to set since it can only have one gateway
- Add get_default_ipv4_route and get_default_ipv6_routes
- Add remove_default_ipv4_route and clear_default_ipv6_routes

### CODE REVIEW

**Comment by jD91mZM2** [comment]
I think `update` is fine, but there's no reason to have it take a closure :)

**Comment by whitequark** [comment]
There is--if you just give away a `&mut` there is no way to maintain any internal invariants. There aren't any right now but that is likely wrong.

**Comment by jD91mZM2** [comment]
What exactly do you mean by internal invariants?

**Comment by whitequark** [comment]
Not every route is legal. For example, no multicast addresses should be added to the route table. This isn't checked currently, but should be, and it can't be checked with `get_mut`.

**Comment by jD91mZM2** [comment]
Aaaaaah okay, that explains it. I'll implement your add_route idea

**Comment by ProgVal** [comment]
The `update` method is inspired from the one in `ethernet.rs`, so you might want to change it as well.

**Comment by jD91mZM2** [comment]
There wasn't much to do in `ethernet.rs` unfortunately since a ManagedSlice cannot be pushed/removed to.

File: src/iface/ethernet.rs [hunk 1]
```diff
@@ -267,13 +267,16 @@ impl<'b, 'c, 'e, DeviceT> Interface<'b, 'c, 'e, DeviceT>
...
-    /// Update the IP addresses of the interface.
+    /// Replace the set of IP addresses.
     ///
     /// # Panics
     /// This function panics if any of the addresses are not unicast.
-    pub fn update_ip_addrs<F: FnOnce(&mut ManagedSlice<'c, IpCidr>)>(&mut self, f: F) {
-        f(&mut self.inner.ip_addrs);
-        InterfaceInner::check_ip_addrs(&self.inner.ip_addrs)
+    pub fn set_ip_addrs<S>(&mut self, ips: S)
+        where S: Into<ManagedSlice<'c, IpCidr>>
```
Comment (whitequark) [inline src/iface/ethernet.rs]: The purpose of this method is to be able to update IP addresses without assigning a different ManagedSlice. It's for memory-constrained devices without an allocator.

Comment (jD91mZM2) [reply src/iface/ethernet.rs]: Well you still can't append to it. Should I just make a way to set an ip based on index?

File: src/iface/route.rs [hunk 2]
```diff
@@ -195,9 +248,7 @@ mod test {
             via_router: ADDR_1A.into(),
             preferred_until: None, expires_at: None,
         };
-        routes.update(|storage| {
-            storage.insert(cidr_1().into(), route).unwrap();
```
Comment (whitequark) [inline src/iface/route.rs]: So what does this achieve? One line less in route setup code? I'm not convinced it's worth adding so many functions in the public API (and you don't even validate input in some of them).

Comment (jD91mZM2) [reply src/iface/route.rs]: Well, imagine you want to fetch the default gateway with the current code:
```Rust
let mut gateway = None;
routes.update(|storage| {
     let cidr = IpCidr::new(IpAddress::v4(0, 0, 0, 0), 0);
    gateway = Some(storage.get(&cidr).clone());
});
let gateway = gateway.unwrap();
```

Comment (jD91mZM2) [reply src/iface/route.rs]: See the [redox netstack code after this PR](https://gitlab.redox-os.org/jD91mZM2/netstack/commit/f9d6f5d87f1f5903d301c6205b9b5637c8ad1512).

(Your existing code doesn't validate input either btw)

File: src/iface/route.rs [hunk 3]
```diff
@@ -74,16 +74,36 @@ impl<'a> Routes<'a> {
...
+    ///
+    /// On success, return the previous route, if any.
+    pub fn add_route(&mut self, cidr: IpCidr, route: Route) -> Result<Option<Route>> {
+        // TODO: Validate
+        match self.storage.insert(cidr, route) {
+            Ok(route) => Ok(route),
+            Err((_cidr, _route)) => Err(Error::Exhausted)
+        }
+    }
+    /// Returns a route registered on cidr
```
Comment (whitequark) [inline src/iface/route.rs]: This docstring isn't informative at all. What does "registered on cidr" mean? It's not clear from the documentation whether this function would only return a route that is an exact match or any route matching the argument.

Also, "CIDR" should be capitalized, it's an acronym.

Also, there's a missing newline.

File: src/iface/route.rs [hunk 4]
```diff
@@ -74,16 +74,36 @@ impl<'a> Routes<'a> {
...
+    pub fn get_route(&mut self, cidr: &IpCidr) -> Option<&Route> {
+        self.storage.get(cidr)
+    }
+    /// Returns a mutable reference to the route registered by cidr
```
Comment (whitequark) [inline src/iface/route.rs]: Same as above.

File: src/iface/route.rs [hunk 5]
```diff
@@ -74,16 +74,36 @@ impl<'a> Routes<'a> {
...
+    pub fn get_route_mut(&mut self, cidr: &IpCidr) -> Option<&mut Route> {
+        self.storage.get_mut(cidr)
+    }
+    /// Delete a route.
+    ///
+    /// On success, returns the value it deleted, if any.
```
Comment (whitequark) [inline src/iface/route.rs]: `returns the route it deleted, if any`

File: src/iface/route.rs [hunk 6]
```diff
@@ -74,16 +74,36 @@ impl<'a> Routes<'a> {
...
+    pub fn del_route(&mut self, cidr: &IpCidr) -> Option<Route> {
```
Comment (whitequark) [inline src/iface/route.rs]: This function should be called `delete_route` (or `remove_route`; not `del_route`).

File: src/iface/route.rs [hunk 7]
```diff
@@ -74,16 +74,36 @@ impl<'a> Routes<'a> {
         Routes { storage }
     }
 
-    /// Update the routes of this node.
-    pub fn update<F: FnOnce(&mut ManagedMap<'a, IpCidr, Route>)>(&mut self, f: F) {
-        f(&mut self.storage);
+    /// Add a new route.
+    ///
+    /// On success, return the previous route, if any.
+    pub fn add_route(&mut self, cidr: IpCidr, route: Route) -> Result<Option<Route>> {
```
Comment (whitequark) [inline src/iface/route.rs]: The naming of this function is inconsistent with `set_*_default_route`. Either this should be called `set_route`, or those functions should be called `add_*_default_route`. They fundamentally do the same thing.

File: src/iface/route.rs [hunk 8]
```diff
@@ -74,16 +74,36 @@ impl<'a> Routes<'a> {
...
     }
 
-    /// Add a default ipv4 gateway (ie. "ip route add 0.0.0.0/0 via `gateway`").
+    /// Set the default ipv4 gateway (ie. "ip route add 0.0.0.0/0 via `gateway`").
     ///
     /// On success, returns the previous default route, if any.
     #[cfg(feature = "proto-ipv4")]
-    pub fn add_default_ipv4_route(&mut self, gateway: Ipv4Address) -> Result<Option<Route>> {
+    pub fn set_default_ipv4_route(&mut self, gateway: Ipv4Address) -> Result<Option<Route>> {
         let cidr = IpCidr::new(IpAddress::v4(0, 0, 0, 0), 0);
```
Comment (whitequark) [inline src/iface/route.rs]: This function should use `self.add_route`.

File: src/iface/route.rs [hunk 9]
```diff
@@ -74,16 +74,36 @@ impl<'a> Routes<'a> {
...
+    }
+    /// Delete a route.
+    ///
+    /// On success, returns the value it deleted, if any.
+    pub fn del_route(&mut self, cidr: &IpCidr) -> Option<Route> {
+        self.storage.remove(&cidr)
     }
 
-    /// Add a default ipv4 gateway (ie. "ip route add 0.0.0.0/0 via `gateway`").
+    /// Set the default ipv4 gateway (ie. "ip route add 0.0.0.0/0 via `gateway`").
```
Comment (whitequark) [inline src/iface/route.rs]: `IPv4 gateway`

File: src/iface/route.rs [hunk 10]
```diff
@@ -92,6 +112,22 @@ impl<'a> Routes<'a> {
         }
     }
 
+    /// Get the current default ipv4 gateway, if any
```
Comment (whitequark) [inline src/iface/route.rs]: `IPv4 gateway`

File: src/iface/route.rs [hunk 11]
```diff
@@ -92,6 +112,22 @@ impl<'a> Routes<'a> {
...
+    #[cfg(feature = "proto-ipv4")]
+    pub fn get_default_ipv4_route(&mut self) -> Option<&Route> {
+        let cidr = IpCidr::new(IpAddress::v4(0, 0, 0, 0), 0);
```
Comment (whitequark) [inline src/iface/route.rs]: This function should use `self.get_route`.

File: src/iface/route.rs [hunk 12]
```diff
@@ -92,6 +112,22 @@ impl<'a> Routes<'a> {
...
+        self.storage.get(&cidr)
+    }
+
+    /// Remove default ipv4 gateway.
+    ///
+    /// Returns the removed route, if any.
+    #[cfg(feature = "proto-ipv4")]
+    pub fn remove_default_ipv4_route(&mut self) -> Option<Route> {
```
Comment (whitequark) [inline src/iface/route.rs]: Once again, inconsistent naming with `del_route`.

File: src/iface/route.rs [hunk 13]
```diff
@@ -92,6 +112,22 @@ impl<'a> Routes<'a> {
...
     }
 
+    /// Get the current default ipv4 gateway, if any
+    #[cfg(feature = "proto-ipv4")]
+    pub fn get_default_ipv4_route(&mut self) -> Option<&Route> {
+        let cidr = IpCidr::new(IpAddress::v4(0, 0, 0, 0), 0);
+        self.storage.get(&cidr)
+    }
+
+    /// Remove default ipv4 gateway.
```
Comment (whitequark) [inline src/iface/route.rs]: `IPv4 gateway`.

File: src/iface/route.rs [hunk 14]
```diff
@@ -105,6 +141,23 @@ impl<'a> Routes<'a> {
         }
     }
 
+    /// Get the current default ipv6 gateways, if any
```
Comment (whitequark) [inline src/iface/route.rs]: `IPv6 gateways`

File: src/iface/route.rs [hunk 15]
```diff
@@ -105,6 +141,23 @@ impl<'a> Routes<'a> {
...
+    #[cfg(feature = "proto-ipv6")]
+    pub fn get_default_ipv6_routes(&mut self) -> impl Iterator<Item = &Route> {
+        let cidr = IpCidr::new(IpAddress::v6(0, 0, 0, 0, 0, 0, 0, 0), 0);
+        // TODO: Support multiple default ipv6 gateways
```
Comment (whitequark) [inline src/iface/route.rs]: Why is this a TODO if you already return an iterator?

File: src/iface/route.rs [hunk 16]
```diff
@@ -105,6 +141,23 @@ impl<'a> Routes<'a> {
...
+        self.storage.get(&cidr).into_iter()
+    }
+
+    /// Clear default ipv6 gateways.
```
Comment (whitequark) [inline src/iface/route.rs]: `IPv6 gateways`

File: src/iface/route.rs [hunk 17]
```diff
@@ -105,6 +141,23 @@ impl<'a> Routes<'a> {
...
+    ///
+    /// On success, returns the cleared routes, if any.
```
Comment (whitequark) [inline src/iface/route.rs]: This function returns exactly one route.

File: src/iface/route.rs [hunk 18]
```diff
@@ -105,6 +141,23 @@ impl<'a> Routes<'a> {
...
+    #[cfg(feature = "proto-ipv6")]
+    pub fn clear_default_ipv6_routes(&mut self) -> Option<Route> {
+        let cidr = IpCidr::new(IpAddress::v6(0, 0, 0, 0, 0, 0, 0, 0), 0);
+        self.storage.remove(&cidr)
```
Comment (whitequark) [inline src/iface/route.rs]: This function should use `self.remove_route`.

**Comment by jD91mZM2** [comment]
Since you previously said you don't think this is needed, I'll just close this because I'm lazy :stuck_out_tongue:. Better now than later when I put in more time into something that isn't even desired.

**Pull request closed without merging**